packages/block-editor/src/components/media-placeholder/index.js

```diff
@@ -116,14 +116,38 @@ export class MediaPlaceholder extends Component {
 			multiple,
 			onError,
 			onSelect,
-			value = [],
 		} = this.props;
 		let setMedia;
 		if ( multiple ) {
 			if ( addToGallery ) {
-				const currentValue = value;
+				// To allow changes to a gallery to be made while uploads are in progress
+				// (including trigging multiple upload groups and removing already in place images),
+				// we must be able to add newMedia based on the current value of the Gallery
+				// whenever the setMedia function runs. Since the setMedia function runs multiple
+				// times per upload group and is passed newMedia containing every item in its
+				// group each time, we must also filter out whatever this upload group had
+				// previously returned to the gallery before adding and returning the image
+				// array with replacement newMedia values.
+
+				// Define an array to store urls from newMedia between subsequent function calls.
+				let lastMediaPassed = [];
 				setMedia = ( newMedia ) => {
-					onSelect( currentValue.concat( newMedia ) );
+					// Get the current array of images on the Gallery.
+					const currentMedia = this.props.value || [];
+					// Remove any images this upload group is responsible for (lastMediaPassed).
+					// Their replacements are contained in newMedia.
+					const filteredMedia = currentMedia.filter( ( item ) => {
+						// Compare via .includes since gallery adds a ratio to the url after returned from this function.
+						return ! lastMediaPassed.some( ( temporaryMediaSlug ) => item.url.includes( temporaryMediaSlug ) );
```

Comment:
This comparison relies on WordPress using a specific format for the filenames of different image sizes. Is there any possibility a plugin could change this format?
Would it be possible for lastMediaPassed be an object that contains id and URL and when the id is available we perform the comparison by id instead of URL to avoid the need to rely on WordPress images with the same id have a common URL prefix?

Reply:
While I am not sure about the plugins changing the format, I give a huge YES to adding the check by Id if it is available.  This should force the url prefix check to only happen on transient images that have not been fully uploaded.

After adding that Id check, the only edge-case I can think of is if `image-name-something` and `image-name` being uploaded simultaneously in different upload groups. Even then, this would be extremely rare and would only temporarily get rid of a transient placeholder that would still be re-added (or inevitably replaced by the fully uploaded image once complete).